hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/DroolsRulesEngineImpl.java

```diff
@@ -147,15 +160,33 @@ public void fire() {
 
             batchData.clear();
 
-            if (log.isTraceEnabled()) {
-                log.trace("Drools session dumping before firing: ");
+
+            TreeSet<Event> batchEvent = new TreeSet<Event>(pendingEvent);
+            Event previousEvent = null;
+
+            pendingEvent.clear();
+
+            for (Event event : batchEvent) {
+                if (null == previousEvent
+                        || (null != event.getDataId() && !event.getDataId().equals(previousEvent.getDataId()))) {
+                    kSession.insert(event);
+                    previousEvent = event;
+                } else {
+                    pendingEvent.add(event);
+                    log.debugf("Deferring more recent %1$s until older %2$s is processed", event, previousEvent);
+                }
+            }
+
+            if (log.isDebugEnabled()) {
+                log.debugf("Firing cycle [%s] - with these facts: ", fireCycle);
                 for (FactHandle fact : kSession.getFactHandles()) {
                     Object o = kSession.getObject(fact);
-                    log.trace("Fact: " + o.toString());
+                    log.debugf("Fact:  %s", o.toString());
```

Comment:
I've moved to debug as I missed it while debugging some scenario.
I guess it is useful to have it on debug as we can have all the info to debug external executions (thinking on some reported error or so), having this info in trace forces to increase a lot the logging output for debugging and I've spent a lot of time just filtering/cutting/chopping the log files to get it.